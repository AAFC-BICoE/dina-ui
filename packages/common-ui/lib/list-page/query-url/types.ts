export interface SimpleQueryGroup {
  // Conjunction (AND/OR)
  c: string;

  // Props: Items inside the group
  p: SimpleQueryRow[];
}

export interface SimpleQueryRow {
  // Field
  f: string;

  // Operator
  o: string;

  // Value
  v: string;

  // Type
  t: string;

  // Dynamic field identifier (UUID for example if managed attribute)
  d?: string;
}
