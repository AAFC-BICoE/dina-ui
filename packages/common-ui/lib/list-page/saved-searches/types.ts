import { JsonTree } from "react-awesome-query-builder";

/**
 * Since saved searches are stored as JSON this is the structure to be used.
 */
export interface SavedSearchStructure {
  [indexName: string]: {
    [savedSearchName: string]: SingleSavedSearch;
  };
}

export interface SingleSavedSearch {
  // This field is not saved in the JSON.
  savedSearchName?: string;
  default: boolean;
  queryTree?: JsonTree;
}
