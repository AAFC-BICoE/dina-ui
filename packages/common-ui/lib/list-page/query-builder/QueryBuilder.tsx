import { LoadingSpinner } from "../..";
import { Query, Builder, Utils, JsonTree } from "react-awesome-query-builder";
import { useCallback } from "react";
import {
  Config,
  ImmutableTree,
  BuilderProps
} from "react-awesome-query-builder";
import React from "react";
import { Button } from "react-bootstrap";
import { SavedSearch } from "../saved-searches/SavedSearch";
import { DinaMessage } from "../../../../dina-ui/intl/dina-ui-intl";

interface QueryBuilderProps {
  /**
   * Index name being used for the QueryPage.
   */
  indexName: string;

  /**
   * Query Builder Configuration.
   */
  queryBuilderConfig?: Config;

  /**
   * Query Builder local tree.
   */
  queryBuilderTree?: ImmutableTree;

  /**
   * Callback function to indicate that the query builder has changed.
   */
  setQueryBuilderTree: (newTree: ImmutableTree) => void;

  /**
   * Callback function to indicate the form was submitted.
   */
  onSubmit: () => void;

  /**
   * Callback to indicate the query builder was reset.
   */
  onReset: () => void;
}

function QueryBuilder({
  indexName,
  queryBuilderConfig,
  queryBuilderTree,
  setQueryBuilderTree,
  onSubmit,
  onReset
}: QueryBuilderProps) {
  const onChange = useCallback((immutableTree: ImmutableTree) => {
    setQueryBuilderTree(immutableTree);
  }, []);

  const renderBuilder = useCallback(
    (props: BuilderProps) => (
      <div className="query-builder-container">
        <div className="query-builder qb-lite">
          <Builder {...props} />
        </div>
      </div>
    ),
    []
  );

  // Display loading spinner when performing request for the index.
  if (!queryBuilderTree || !queryBuilderConfig) {
    return <LoadingSpinner loading={true} />;
  }

  return (
    <>
      <label
        style={{
          fontSize: 20,
          fontFamily: "sans-serif",
          fontWeight: "bold"
        }}
        className="mb-2"
      >
        <DinaMessage id="search" />
      </label>
      <SavedSearch
        indexName={indexName}
        queryBuilderTree={queryBuilderTree}
        setQueryBuilderTree={setQueryBuilderTree}
        queryBuilderConfig={queryBuilderConfig}
        performSubmit={onSubmit}
      />
      <Query
        {...queryBuilderConfig}
        value={queryBuilderTree}
        onChange={onChange}
        renderBuilder={renderBuilder}
      />
      <div className="mt-2">
        <Button onClick={onSubmit} className="me-2">
          Search
        </Button>
        <Button onClick={onReset} variant="secondary">
          Reset
        </Button>
      </div>
    </>
  );
}

/**
 * Default query tree contains an empty rule so it's one less the step the user needs to perform.
 *
 * This is the default when first loading the page and resetting the query builder.
 */
export function defaultQueryTree(): ImmutableTree {
  const groupId = "8c6dc2c8-4070-48ce-b700-13a931f9ebaf";
  const ruleId = "f76a54f6-0112-4ac9-b2a1-f6dced58b3d6";

  return Utils.loadTree({
    id: groupId,
    type: "group",
    children1: {
      "f76a54f6-0112-4ac9-b2a1-f6dced58b3d6": {
        type: "rule",
        id: ruleId,
        properties: {
          field: null,
          operator: null,
          value: [],
          valueSrc: [],
          valueError: []
        },
        path: [groupId, ruleId]
      }
    },
    properties: { conjunction: "AND" },
    path: [groupId]
  } as JsonTree);
}

/**
 * Generate an empty tree with no rules. Used for special cases like custom elastic search
 * view queries where the query builder is not needed.
 */
export function emptyQueryTree(): ImmutableTree {
  return Utils.loadTree({
    id: Utils.uuid(),
    type: "group",
    children1: {}
  } as JsonTree);
}

/**
 * Currently utilized for the custom view queries to show associated records.
 *
 * @param uuid UUID to search for.
 * @param path Relationship path to search against.
 * @returns QueryBuilder tree which will be translated into elastic search query.
 */
export function generateUUIDTree(uuid: string, path: string): JsonTree {
  return {
    id: Utils.uuid(),
    type: "group",
    children1: {
      [Utils.uuid()]: {
        type: "rule",
        properties: {
          field: path,
          operator: "uuid",
          value: [uuid]
        }
      }
    }
  };
}

export const QueryBuilderMemo = React.memo(QueryBuilder);
