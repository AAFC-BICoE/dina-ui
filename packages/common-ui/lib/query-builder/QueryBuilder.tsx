import { useMemo } from "react";
import useSWR from "swr";
import { DinaForm, useApiClient } from "..";
import { QueryRow } from "./QueryRow";
import { v4 as uuidv4 } from "uuid";

export function QueryBuilder() {
  const { apiClient } = useApiClient();

  async function fetchQueryFieldsByIndex(indexName) {
    const resp = await apiClient.axios.get("search-api/search/mapping", {
      params: { indexName }
    });

    const result: [{}] = [{}];

    Object.keys(resp.data)
      .filter(key => key.includes("data.attributes."))
      .map(key => {
        const fieldNameLabel = key.substring(
          "data.attributes.".length,
          key.lastIndexOf(".")
        );
        result.push({
          label: fieldNameLabel,
          value: fieldNameLabel,
          type: resp.data?.[key]
        });
      });

    return result;
  }

  // Invalidate the query cache on query change, don't use SWR's built-in cache:
  const cacheId = useMemo(() => uuidv4(), []);

  const {
    data,
    error,
    isValidating: loading
  } = useSWR(["dina_material_sample_index", cacheId], fetchQueryFieldsByIndex, {
    shouldRetryOnError: true,
    revalidateOnFocus: false,
    revalidateOnReconnect: false
  });

  if (loading || error) return <></>;

  return (
    <DinaForm initialValues={{}}>
      <QueryRow queryRowProps={data} />
    </DinaForm>
  );
}
