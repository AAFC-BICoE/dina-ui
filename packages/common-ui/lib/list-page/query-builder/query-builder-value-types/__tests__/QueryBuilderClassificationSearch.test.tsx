import { mountWithAppContext } from "common-ui/lib/test-util/mock-app-context";
import { DinaForm } from "common-ui/lib/formik-connected/DinaForm";
import { QueryBuilderContextProvider } from "../../QueryBuilder";
import { noop } from "lodash";
import userEvent from "@testing-library/user-event";
import { Vocabulary } from "packages/dina-ui/types/collection-api";
import QueryRowClassificationSearch, {
  SUPPORTED_CLASSIFICATION_OPERATORS,
  transformClassificationToDSL
} from "../QueryBuilderClassificationSearch";

export const TEST_CLASSIFICATIONS: Vocabulary = {
  id: "taxonomicRank",
  type: "vocabulary",
  vocabularyElements: [
    {
      key: "kingdom",
      name: "kingdom",
      multilingualTitle: {
        titles: [
          {
            lang: "en",
            title: "kingdom"
          },
          {
            lang: "fr",
            title: "r√®gne"
          }
        ]
      }
    }
  ]
};

const mockGet = jest.fn<any, any>(async (path) => {
  switch (path) {
    case "collection-api/vocabulary2/taxonomicRank":
      return { data: TEST_CLASSIFICATIONS };
  }
});

const apiContext: any = {
  apiClient: { get: mockGet }
};

describe("QueryBuilderClassificationSearch", () => {
  describe("QueryBuilderClassificationSearch Component", () => {
    it("Display field if match type is equals", async () => {
      // This test will just ensure the layout does not change unexpectedly.
      // Any changes to the layout, the snapshots will need to be updated.
      const textSearchEquals = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider
            value={{ performSubmit: noop, groups: [] }}
          >
            <QueryRowClassificationSearch
              value='{"searchValue":"","selectedOperator":"exactMatch","selectedClassificationRank":"class"}'
              setValue={jest.fn}
              isInColumnSelector={false}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        {
          apiContext
        }
      );

      // Expect a snapshot with the text field being displayed.
      expect(textSearchEquals.asFragment()).toMatchSnapshot(
        "Expect text field to be displayed since match type is equals"
      );

      const textSearchEmpty = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider
            value={{ performSubmit: noop, groups: [] }}
          >
            <QueryRowClassificationSearch
              value='{"searchValue":"","selectedOperator":"empty","selectedClassificationRank":"class"}'
              setValue={jest.fn}
              isInColumnSelector={false}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        {
          apiContext
        }
      );

      // Expect a snapshot without the text field being displayed.
      expect(textSearchEmpty.asFragment()).toMatchSnapshot(
        "Expect text field not to be displayed since the match type is not equals"
      );
    });

    it("Should call performSubmit on enter key press in textfield", async () => {
      const mockPerformSubmit = jest.fn();
      const wrapper = mountWithAppContext(
        <DinaForm initialValues={{}}>
          <QueryBuilderContextProvider
            value={{ performSubmit: mockPerformSubmit, groups: [] }}
          >
            <QueryRowClassificationSearch
              value='{"searchValue":"","selectedOperator":"exactMatch","selectedClassificationRank":"class"}'
              setValue={jest.fn}
              isInColumnSelector={false}
            />
          </QueryBuilderContextProvider>
        </DinaForm>,
        {
          apiContext
        }
      );

      // Find the text field element
      const textField = wrapper.getByRole("textbox");

      // Expect performSubmit to not be called yet.
      expect(mockPerformSubmit).toHaveBeenCalledTimes(0);

      // Simulate user typing "enter" key
      userEvent.type(textField, "{enter}");

      // Expect performSubmit to be called once
      expect(mockPerformSubmit).toHaveBeenCalledTimes(1);
    });
  });

  describe("transformClassificationToDSL function", () => {
    // Loop through all the possible operators and generate a snapshot for each.
    SUPPORTED_CLASSIFICATION_OPERATORS.forEach((operator) => {
      test(`Operation: ${operator}`, async () => {
        // Attribute level
        expect(
          transformClassificationToDSL({
            value: `{"selectedClassificationRank":"kingdom","selectedOperator":"${operator}","searchValue":"Animalia"}`,
            fieldPath:
              "included.attributes.targetOrganismPrimaryClassification",
            fieldInfo: {
              path: "included.attributes.targetOrganismPrimaryClassification",
              distinctTerm: false,
              hideField: false,
              keywordMultiFieldSupport: false,
              containsSupport: false,
              endsWithSupport: false,
              keywordNumericSupport: false,
              optimizedPrefix: false,
              label: "Target Organism Primary Classification",
              type: "classification",
              value: ""
            },
            operation: "noOperator",
            queryType: "classification"
          })
        ).toMatchSnapshot();

        // Parent level test
        expect(
          transformClassificationToDSL({
            value: `{"selectedClassificationRank":"kingdom","selectedOperator":"${operator}","searchValue":"Animalia"}`,
            fieldPath:
              "included.attributes.targetOrganismPrimaryClassification",
            fieldInfo: {
              path: "included.attributes.targetOrganismPrimaryClassification",
              parentName: "parentMaterialSample",
              parentType: "material-sample",
              distinctTerm: false,
              hideField: false,
              keywordMultiFieldSupport: false,
              containsSupport: false,
              endsWithSupport: false,
              keywordNumericSupport: false,
              optimizedPrefix: false,
              label: "Target Organism Primary Classification",
              type: "classification",
              value: ""
            },
            operation: "noOperator",
            queryType: "classification"
          })
        ).toMatchSnapshot();
      });
    });
  });
});
