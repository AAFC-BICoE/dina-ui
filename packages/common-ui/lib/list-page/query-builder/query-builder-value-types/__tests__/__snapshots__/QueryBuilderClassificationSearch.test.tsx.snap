// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilderClassificationSearch QueryBuilderClassificationSearch Component Display field if match type is equals: Expect text field not to be displayed since the match type is not equals 1`] = `
<DocumentFragment>
  <div>
    <form
      action="#"
    >
      <div
        class="error-viewer"
      />
      <div
        class="row"
      >
        <div
          class="col me-1 ms-2 ps-0 css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-4-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class="css-13cymwt-control"
          >
            <div
              class="css-hlgwow"
            >
              <div
                class="css-1jqq78o-placeholder"
                id="react-select-4-placeholder"
              >
                Select classification rank...
              </div>
              <div
                class="css-19bb58m"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-describedby="react-select-4-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-4-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-1wy0on6"
            >
              <div
                aria-hidden="true"
                class="css-pogzpp-loadingIndicator"
              >
                <span
                  class="css-1icxkl9-LoadingDot"
                />
                <span
                  class="css-1oazjx7-LoadingDot"
                />
                <span
                  class="css-1wonz24-LoadingDot"
                />
              </div>
              <span
                class="css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col me-1 ps-0 css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-5-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class="css-13cymwt-control"
          >
            <div
              class="css-hlgwow"
            >
              <div
                class="css-1dimb5e-singleValue"
              >
                Empty
              </div>
              <div
                class="css-19bb58m"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-5-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-1wy0on6"
            >
              <span
                class="css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`QueryBuilderClassificationSearch QueryBuilderClassificationSearch Component Display field if match type is equals: Expect text field to be displayed since match type is equals 1`] = `
<DocumentFragment>
  <div>
    <form
      action="#"
    >
      <div
        class="error-viewer"
      />
      <div
        class="row"
      >
        <div
          class="col me-1 ms-2 ps-0 css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class="css-13cymwt-control"
          >
            <div
              class="css-hlgwow"
            >
              <div
                class="css-1jqq78o-placeholder"
                id="react-select-2-placeholder"
              >
                Select classification rank...
              </div>
              <div
                class="css-19bb58m"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-describedby="react-select-2-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-2-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-1wy0on6"
            >
              <div
                aria-hidden="true"
                class="css-pogzpp-loadingIndicator"
              >
                <span
                  class="css-1icxkl9-LoadingDot"
                />
                <span
                  class="css-1oazjx7-LoadingDot"
                />
                <span
                  class="css-1wonz24-LoadingDot"
                />
              </div>
              <span
                class="css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col me-1 ps-0 css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-3-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class="css-13cymwt-control"
          >
            <div
              class="css-hlgwow"
            >
              <div
                class="css-1dimb5e-singleValue"
              >
                Exact match
              </div>
              <div
                class="css-19bb58m"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-3-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-1wy0on6"
            >
              <span
                class="css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <input
          class="col form-control me-3"
          placeholder="Enter text search value..."
          type="text"
          value=""
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: empty 1`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "bool": Object {
          "must_not": Object {
            "exists": Object {
              "field": "included.attributes.targetOrganismPrimaryClassification.kingdom",
            },
          },
        },
      },
      Object {
        "bool": Object {
          "must": Object {
            "term": Object {
              "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": "",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: empty 2`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "bool": Object {
          "should": Array [
            Object {
              "bool": Object {
                "must_not": Object {
                  "nested": Object {
                    "path": "included",
                    "query": Object {
                      "bool": Object {
                        "must": Array [
                          Object {
                            "exists": Object {
                              "field": "included.attributes.targetOrganismPrimaryClassification.kingdom",
                            },
                          },
                          Object {
                            "term": Object {
                              "included.type": "material-sample",
                            },
                          },
                        ],
                      },
                    },
                  },
                },
              },
            },
            Object {
              "nested": Object {
                "path": "included",
                "query": Object {
                  "bool": Object {
                    "must": Array [
                      Object {
                        "term": Object {
                          "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": "",
                        },
                      },
                      Object {
                        "term": Object {
                          "included.type": "material-sample",
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      Object {
        "bool": Object {
          "must_not": Object {
            "term": Object {
              "included.type": "material-sample",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: exactMatch 1`] = `
Object {
  "term": Object {
    "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": "Animalia",
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: exactMatch 2`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": "Animalia",
            },
          },
          Object {
            "term": Object {
              "included.type": "material-sample",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: in 1`] = `
Object {
  "bool": Object {
    "must": Object {
      "bool": Object {
        "minimum_should_match": 1,
        "should": Array [
          Object {
            "term": Object {
              "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": Object {
                "case_insensitive": true,
                "value": "Animalia",
              },
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: in 2`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "bool": Object {
              "must": Object {
                "bool": Object {
                  "minimum_should_match": 1,
                  "should": Array [
                    Object {
                      "term": Object {
                        "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": Object {
                          "case_insensitive": true,
                          "value": "Animalia",
                        },
                      },
                    },
                  ],
                },
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "material-sample",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: notEmpty 1`] = `
Object {
  "bool": Object {
    "must": Object {
      "exists": Object {
        "field": "included.attributes.targetOrganismPrimaryClassification.kingdom",
      },
    },
    "must_not": Object {
      "term": Object {
        "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": "",
      },
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: notEmpty 2`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "included.type": "material-sample",
            },
          },
          Object {
            "exists": Object {
              "field": "included.attributes.targetOrganismPrimaryClassification.kingdom",
            },
          },
        ],
        "must_not": Object {
          "term": Object {
            "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": "",
          },
        },
      },
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: notEquals 1`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "bool": Object {
          "must_not": Object {
            "term": Object {
              "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": "Animalia",
            },
          },
        },
      },
      Object {
        "bool": Object {
          "must_not": Object {
            "exists": Object {
              "field": "included.attributes.targetOrganismPrimaryClassification.kingdom",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: notEquals 2`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "nested": Object {
          "path": "included",
          "query": Object {
            "bool": Object {
              "must": Object {
                "term": Object {
                  "included.type": "material-sample",
                },
              },
              "must_not": Object {
                "term": Object {
                  "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": "Animalia",
                },
              },
            },
          },
        },
      },
      Object {
        "nested": Object {
          "path": "included",
          "query": Object {
            "bool": Object {
              "must": Object {
                "term": Object {
                  "included.type": "material-sample",
                },
              },
              "must_not": Object {
                "exists": Object {
                  "field": "included.attributes.targetOrganismPrimaryClassification.kingdom",
                },
              },
            },
          },
        },
      },
      Object {
        "bool": Object {
          "must_not": Object {
            "term": Object {
              "included.type": "material-sample",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: notIn 1`] = `
Object {
  "bool": Object {
    "must_not": Object {
      "bool": Object {
        "minimum_should_match": 1,
        "should": Array [
          Object {
            "term": Object {
              "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": Object {
                "case_insensitive": true,
                "value": "Animalia",
              },
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: notIn 2`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "bool": Object {
              "must_not": Object {
                "bool": Object {
                  "minimum_should_match": 1,
                  "should": Array [
                    Object {
                      "term": Object {
                        "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": Object {
                          "case_insensitive": true,
                          "value": "Animalia",
                        },
                      },
                    },
                  ],
                },
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "material-sample",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: startsWith 1`] = `
Object {
  "prefix": Object {
    "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": Object {
      "case_insensitive": true,
      "value": "Animalia",
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: startsWith 2`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "prefix": Object {
              "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": Object {
                "case_insensitive": true,
                "value": "Animalia",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "material-sample",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: wildcard 1`] = `
Object {
  "wildcard": Object {
    "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": Object {
      "case_insensitive": true,
      "value": "*Animalia*",
    },
  },
}
`;

exports[`QueryBuilderClassificationSearch transformClassificationToDSL function Operation: wildcard 2`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "wildcard": Object {
              "included.attributes.targetOrganismPrimaryClassification.kingdom.keyword": Object {
                "case_insensitive": true,
                "value": "*Animalia*",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "material-sample",
            },
          },
        ],
      },
    },
  },
}
`;
